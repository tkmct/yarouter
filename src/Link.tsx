import * as React from 'react'
import LocationContext from './locationContext'

interface Props {
  children: React.ReactChildren
  to: string
}

export default function({ children, to }: Props): JSX.Element {
  const { history } = React.useContext(LocationContext)
  function onClick(e: React.MouseEvent<HTMLAnchorElement>) {
    e.preventDefault()
    history.push(to)
  }

  return <a onClick={onClick}>{children}</a>
}
