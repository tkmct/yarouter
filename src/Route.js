import { useContext } from 'react'
import { LocationContext } from './Router'

export default function Route({ path, component }) {
  const location = useContext(LocationContext)

  // TODO: extract to util function
  if (path === location.pathname) {
    return component()
  }

  return null
}
