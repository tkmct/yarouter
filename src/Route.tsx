interface Props {
  path: string
  component: () => JSX.Element
  onEnter?: () => any
}

export default function Route({ component, onEnter }: Props) {
  if (onEnter) {
    onEnter()
  }

  return component()
}
