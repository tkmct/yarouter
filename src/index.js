import React, { Component } from 'react'

export default class Router extends Component {
  render() {
    return <div children={<p>Router</p>} />
  }
}
