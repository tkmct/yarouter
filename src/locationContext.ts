import * as React from 'react'
import { History, Location } from 'history'

const initialLocation = {
  pathname: '',
  search: '',
  state: {},
  hash: '',
}

export default React.createContext<{ location: Location; history: History }>({
  location: initialLocation,
  history: ({} as any) as History,
})
