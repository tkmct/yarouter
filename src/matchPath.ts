import * as React from 'react'

export default function matchPath(
  location: { pathname: string },
  routes: Array<React.ReactElement<any>>
): React.ReactElement<any> | null {
  return routes.find(route => route.props.path === location.pathname) || null
}
