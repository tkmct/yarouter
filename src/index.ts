import Router, { LocationContext } from './Router'
import Route from './Route'
import Link from './Link'
import { createBrowserHistory } from 'history'

export { Router, Route, Link, LocationContext, createBrowserHistory }
