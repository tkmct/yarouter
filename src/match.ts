import * as React from 'react'
import { Props as RouteProps } from './Route'
import { trimTrailingSlash } from './util'

type RouteElement = React.ReactElement<RouteProps>

interface MatchProps {
  path: string
  exact?: boolean
}

export function matchExactPath(path: string, currentPath: string): boolean {
  return path === currentPath
}

export function matchPath({ path, exact }: MatchProps, currentPath: string): boolean {
  const trimedCurrentPath = currentPath === '/' ? currentPath : trimTrailingSlash(currentPath)

  if (exact) {
    return matchExactPath(path, trimedCurrentPath)
  }
  return trimedCurrentPath.startsWith(path)
}

function getRoutedComponent(route: RouteElement | null): React.ComponentType<any> | null {
  if (!route) {
    return null
  }

  return route.props.component
}

export default function match(
  { pathname }: { pathname: string },
  routes: RouteElement[] | RouteElement
): React.ComponentType<any> | null {
  if (!Array.isArray(routes)) {
    const routeComponent = matchPath(routes.props, pathname) ? routes : null
    return getRoutedComponent(routeComponent)
  }

  const routeComponent = routes.find(route => matchPath(route.props, pathname)) || null
  return getRoutedComponent(routeComponent)
}
