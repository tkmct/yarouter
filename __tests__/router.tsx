import React from 'react'
import { Router, Route, createBrowserHistory } from '../src'
import { render } from 'react-testing-library'

function Hello() {
  return <p>Hello</p>
}
function World() {
  return <p>World</p>
}

const setupComponent = () => {
  const history = createBrowserHistory()
  const component = (
    <Router history={history}>
      <Route path="/" component={Hello} />
      <Route path="/world" component={World} />
    </Router>
  )

  return { history, component }
}

describe('Test router component', () => {
  test('snapshot', () => {
    const { component } = setupComponent()
    const { container } = render(component)
    expect(container).toMatchSnapshot()
  })

  test('render', () => {
    const { component } = setupComponent()
    const { getByText } = render(component)
    expect(getByText('Hello')).not.toBeNull()
  })
})
