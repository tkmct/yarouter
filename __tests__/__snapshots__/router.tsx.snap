// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test router component snapshot 1`] = `
<div>
  <p>
    Hello
  </p>
  <div
    style="font-family: sans-serif; position: fixed; bottom: 10px; right: 10px; padding: 10px; border-radius: 4px; max-width: 240px; white-space: pre-wrap; box-shadow: lightgrey 1px 1px 2px 1px;"
  >
    {"pathname": "/", "search": "", "hash": ""}
  </div>
</div>
`;
